pkg/controller/controller.go

```diff
@@ -418,75 +433,145 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
+		output, err := deferPhase.Exec(context.Background(), *bp, actionName, *tp)
+		var rf func(*crv1alpha1.ActionSet) error
 		if err != nil {
-			af = func(ras *crv1alpha1.ActionSet) error {
-				ras.Status.State = crv1alpha1.StateFailed
-				ras.Status.Error = crv1alpha1.Error{
+			rf = func(as *crv1alpha1.ActionSet) error {
+				as.Status.State = crv1alpha1.StateFailed
+				as.Status.Error = crv1alpha1.Error{
 					Message: err.Error(),
 				}
+				as.Status.Actions[aIDX].DeferPhase.State = crv1alpha1.StateFailed
 				return nil
 			}
 		} else {
-			af = func(ras *crv1alpha1.ActionSet) error {
-				ras.Status.Actions[aIDX].Artifacts = arts
-				ras.Status.State = crv1alpha1.StateComplete
+			rf = func(as *crv1alpha1.ActionSet) error {
+				as.Status.Actions[aIDX].DeferPhase.State = crv1alpha1.StateComplete
+				as.Status.Actions[aIDX].DeferPhase.Output = output
 				return nil
 			}
 		}
-		// Update ActionSet
-		if aErr := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), ns, name, af); aErr != nil {
-			reason := fmt.Sprintf("ActionSetFailed Action: %s", action.Name)
-			msg := fmt.Sprintf("Failed to update Output Artifacts: %#v:", artTpls)
-			c.logAndErrorEvent(ctx, msg, reason, aErr, as, bp)
-			return nil
+		var msg string
+		if rErr := reconcile.ActionSet(context.Background(), c.crClient.CrV1alpha1(), actionsetNS, actionsetName, rf); rErr != nil {
+			reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
+			msg := fmt.Sprintf("Failed to update defer phase: %#v:", as.Status.Actions[aIDX].DeferPhase)
+			c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
+			return
 		}
+
 		if err != nil {
-			reason := fmt.Sprintf("ActionSetFailed Action: %s", action.Name)
-			msg := "Failed to render output artifacts"
+			reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
+			if msg == "" {
+				msg = fmt.Sprintf("Failed to execute defer phase: %#v:", as.Status.Actions[aIDX].DeferPhase)
+			}
 			c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
+			return
+		}
+
+		c.logAndSuccessEvent(ctx, fmt.Sprintf("Completed deferPhase %s", as.Status.Actions[aIDX].DeferPhase.Name), "Ended deferPhase", as)
+		param.UpdateDeferPhaseParams(context.Background(), tp, output)
+	}
+
+	// Check if output artifacts are present
+	artTpls := as.Status.Actions[aIDX].Artifacts
+	if len(artTpls) == 0 {
+		// No artifacts, set ActionSetStatus to complete
+		if rErr := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), actionsetNS, actionsetName, func(ras *crv1alpha1.ActionSet) error {
+			ras.Status.State = crv1alpha1.StateComplete
 			return nil
+		}); rErr != nil {
+			reason := fmt.Sprintf("ActionSetFailed Action: %s", actionName)
+			msg := fmt.Sprintf("Failed to update ActionSet: %s", actionsetName)
+			c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
 		}
-		return nil
-	})
-	return nil
+		return
+	}
+	// Render the artifacts
+	arts, err := param.RenderArtifacts(artTpls, *tp)
```

Comment:
How do phases render their artifacts? It seems like this is the only place where `param.RenderArtifacts()` is called.

Reply:
all the phases (core and defer) set their artifact in actionset's status, and then all of them are rendered using this call.
Rendering of the artifact is covered in testcase as part of [this PR](https://github.com/kanisterio/kanister/pull/1337).

Reply:
Thanks for the explanation. Can we move line 531 to 582 into a separate function? Looks like they have nothing to do with `executeDeferPhase()`. Then we can also perform the `if deferPhase != nil` conditional earlier, say at line 417.